src/TightBindingApproximation.jl

```diff
@@ -400,83 +401,209 @@ function run!(tba::Algorithm{<:AbstractTBA}, eb::Assignment{<:EnergyBands})
         eb.data[2][i, :] = real(eigenvalues)
     end
 end
+"""
+    abstract type AbstractMethod end
 
+Abstract type for calculation of Berry curvature.
+"""
+abstract type AbstractMethod end
 """
-    BerryCurvature{B<:Union{BrillouinZone, ReciprocalZone}, O} <: Action
+    Fukui <: AbstractMethod
 
-Berry curvature of energy bands with the spirit of a momentum space discretization method by [Fukui et al, JPSJ 74, 1674 (2005)](https://journals.jps.jp/doi/10.1143/JPSJ.74.1674).
+Fukui method to calculate Berry curvature of energy bands. see [Fukui et al, JPSJ 74, 1674 (2005)](https://journals.jps.jp/doi/10.1143/JPSJ.74.1674). Hall conductivity (single band) is given by 
+```math
+\\sigma_{xy} = -{e^2\\over h}\\sum_n c_n, \\nonumber \\\\
+c_n = -{1\\over 2\\pi}\\int_{BZ}{dk_x dk_y Ω_{xy}}, 
+\\Omega_{xy}=(\\nabla\\times {\\bm A})_z,
+A_{x}=i\\langle u_n|\\partial_x|u_n\\rangle.
+```
 """
-struct BerryCurvature{B<:Union{BrillouinZone, ReciprocalZone}, O} <: Action
-    reciprocalspace::B
+struct Fukui <: AbstractMethod
     bands::Vector{Int}
+    abelian::Bool
+end
+@inline Fukui(bands::AbstractVector{Int}; abelian::Bool=true) = Fukui(collect(bands), abelian)
+"""
+    Kubo{O} <: AbstractMethod
+    
+Kubo method to calculate the total Berry curvature of occupied energy bands. The Kubo formula is given by 
+```math
+\\Omega_{ij}(\\bm k)=\\epsilon_{ijl}\\sum_{n}f(\\epsilon_n(\\bm k))b_n^l(\\bm k)=-2{\\rm Im}\\sum_v\\sum_c{V_{vc,i}(\\bm k)V_{cv,j}(\\bm k)\\over [\\omega_c(\\bm k)-\\omega_v(\\bm k)]^2},
+```
+where 
+```math
+ V_{cv,j}={\\langle u_{c\\bm k}|{\\partial H\\over \\partial {\\bm k}_j}|u_{v\\bm k}\\rangle}
+``` 
+v and c subscripts denote valence (occupied) and conduction (unoccupied) bands, respectively.
+Hall conductivity in 2D space is given by 
+```math
+\\sigma_{xy}=-{e^2\\over h}\\int_{BZ}{dk_x dk_y\\over 2\\pi}{\\Omega_{xy}}
+```
+"""
+struct Kubo{O} <: AbstractMethod 
+    μ::Float64
+    options::O
```

Comment:
I think it is better to expand the `options` attribute. After reading the following codes, it seems to me that  this `options` attribute contains only two fields, `dx` and `dy`. It is convenient to use `dx` and `dy` instead as the attributes of `Kubo` directly.

Reply:
I have moved the `options` attribute and added `kx`, `ky` and `d`.